src/directives/repeat.ts

```diff
@@ -41,67 +95,296 @@ export function repeat<T>(
+        insertPartBefore(
+            containerPart, oldParts[oldHead]!, newParts[newTail + 1]);
+        oldHead++;
+        newTail--;
+      } else if (oldKeys[oldTail] === newKeys[newHead]) {
+        // Old tail matches new head; update and move to new head
+        newParts[newHead] = updatePart(oldParts[oldTail]!, newValues[newHead]);
+        insertPartBefore(containerPart, oldParts[oldTail]!, oldParts[oldHead]!);
+        oldTail--;
+        newHead++;
+      } else {
+        if (newKeyToIndexMap! === undefined) {
+          // Lazily generate key-to-index maps, used for removals & moves below
+          newKeyToIndexMap = generateMap(newKeys, newHead, newTail);
+          oldKeyToIndexMap = generateMap(oldKeys, oldHead, oldTail);
         }
-      } else if (currentMarker !== itemPart.startNode) {
-        // Existing part in the wrong position
-        const end = itemPart.endNode.nextSibling!;
-        if (currentMarker !== end) {
-          reparentNodes(container, itemPart.startNode, end, currentMarker);
+        if (!newKeyToIndexMap!.has(oldKeys[oldHead])) {
+          // Old head is no longer in new list; remove
+          removePart(oldParts[oldHead]!);
+          oldHead++;
+        } else if (!newKeyToIndexMap!.has(oldKeys[oldTail])) {
+          // Old tail is no longer in new list; remove
+          removePart(oldParts[oldTail]!);
+          oldTail--;
+        } else {
+          // Any mismatches at this point are due to additions or moves; see if
+          // we have an old part we can reuse and move into place
+          const oldIndex = oldKeyToIndexMap!.get(newKeys[newHead]);
+          const oldPart = oldIndex !== undefined ? oldParts[oldIndex] : null;
+          if (oldPart == null) {
+            // No old part for this value; create a new one and insert it
+            let newPart =
+                createAndInsertPart(containerPart, oldParts[oldHead]!);
+            updatePart(newPart, newValues[newHead]);
+            newParts[newHead] = newPart;
+          } else {
+            // Reuse old part
+            newParts[newHead] = updatePart(oldPart, newValues[newHead]);
+            insertPartBefore(containerPart, oldPart, oldParts[oldHead]!);
+            // This marks the old part as having been used, so that it will be
+            // skipped in the first two checks above
+            oldParts[oldIndex as number] = null;
+          }
+          newHead++;
         }
-      } else {
-        // else part is in the correct position already
-        currentMarker = itemPart.endNode.nextSibling!;
       }
-
-      itemPart.setValue(result);
-      itemPart.commit();
     }
-
-    // Cleanup
-    if (currentMarker !== part.endNode) {
-      removeNodes(container, currentMarker, part.endNode);
-      keyMap.forEach(cleanMap);
+    // Add parts for any remaining new values
+    while (newHead <= newTail) {
+      const newPart = createAndInsertPart(containerPart, oldParts[oldHead]!);
```

Comment:
Can you insert before `newParts[newTail]` here? Using the `oldParts` as anchors is confusing so it'd be nice to only doing it when we have.

Reply:
The `newTail` has already been advanced and so it would actually be `newParts[newTail+1]`.

But, I already described the rules for where to insert in both cases here: https://github.com/Polymer/lit-html/pull/501/files#diff-fa22dc099475a72fdfd85c634894b896R202

Even though we're inserting at the "head", the above is only possible since the old parts are out of the picture at this point in the algorithm, so it would be a special difference that would need to be explained, and I think I'd prefer to keep the rule simple as documented: head goes at `oldParts[oldHead]`, tail goes at `newParts[newTail+1]`